contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractTupleUniqueExactlyOnceKafkaOutputOperator.java

```diff
@@ -0,0 +1,612 @@
+      return true;
+    }
+
+    if (!equals(cachedMsg.first, msg.first)) {
+      return false;
+    }
+
+    return equals(cachedMsg.second, msg.second);
+  }
+
+  /**
+   * 
+   * @param bytes
+   * @param value
+   * @return
+   */
+  protected abstract <M> boolean equals(byte[] bytes, M value);
+
+  /**
+   * get the partition key. for 0.8.1, If a partition key is provided it will
+   * override the key for the purpose of partitioning but will not be stored.
+   * 
+   * @return
+   */
+  protected int getPartitionKey(K key)
+  {
+    if (partitioner != null) {
+      return partitioner.partition(key, partitionNum);
+    }
+
+    if (key != null) {
+      return key.hashCode();
+    }
+
+    //stick to the Kafka partition, so can't use round robbin
+    return 0;
+  }
+
+  /**
+   * setup the configuration for control producer
+   * 
+   * @return
+   */
+  protected ProducerConfig createKafkaControlProducerConfig()
+  {
+    if (controlProducerProperties == null || controlProducerProperties.isEmpty()) {
+      controlProducerProperties = getProducerProperties();
+    }
+
+    Properties prop = new Properties();
+    for (String propString : controlProducerProperties.split(",")) {
+      if (!propString.contains("=")) {
+        continue;
+      }
+      String[] keyVal = StringUtils.trim(propString).split("=");
+      prop.put(StringUtils.trim(keyVal[0]), StringUtils.trim(keyVal[1]));
+    }
+
+    //only support String encoder now, overwrite
+    prop.setProperty("serializer.class", "kafka.serializer.StringEncoder");
+    prop.setProperty("key.serializer.class", "kafka.serializer.StringEncoder");
+
+    Properties configProperties = this.getConfigProperties();
+    configProperties.putAll(prop);
+
+    return new ProducerConfig(configProperties);
+  }
+
+  /**
+   * Tell the operator how to convert a input tuple to a kafka key value pair
+   * 
+   * @param tuple
+   * @return A kafka key value pair.
+   */
+  protected abstract Pair<K, V> tupleToKeyValue(T tuple);
+
+  public kafka.producer.Partitioner getPartitioner()
+  {
+    return partitioner;
+  }
+
+  public void setPartitioner(kafka.producer.Partitioner partitioner)
+  {
+    this.partitioner = partitioner;
+  }
+
+  public String getControlTopic()
+  {
+    return controlTopic;
+  }
+
+  public void setControlTopic(String controlTopic)
+  {
+    this.controlTopic = controlTopic;
+  }
+
+  public String getControlProducerProperties()
+  {
+    return controlProducerProperties;
+  }
+
+  public void setControlProducerProperties(String controlProducerProperties)
+  {
+    this.controlProducerProperties = controlProducerProperties;
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractExactlyOnceKafkaOutputOperator.class);
+
+  /**
+   * This class used to keep the recovery information
+   *
+   */
+  protected static class RecoveryControlInfo
+  {
+    protected static final String SEPERATOR = "#";
```

Comment:
The sub classes might need to use it( for example when over write the logic of AbstractTupleUniqueExactlyOnceKafkaOutputOperator.RecoveryControlInfo.fromString(String) ). modified by private would have a lot of pain in this case.